core/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -267,376 +302,264 @@ protected void doRun() throws Exception {
-                return;
-            }
-        }
 
-        // if there is no shard, create it
-        if (!indexService.hasShard(shardId)) {
-            if (failedShards.containsKey(shardRouting.shardId())) {
-                if (nodes.getMasterNode() != null) {
-                    String message = "master " + nodes.getMasterNode() + " marked shard as initializing, but shard is marked as failed, resend shard failure";
-                    logger.trace("[{}] re-sending failed shard [{}], reason [{}]", shardRouting.shardId(), shardRouting, message);
-                    shardStateAction.shardFailed(shardRouting, shardRouting, message, null, SHARD_STATE_ACTION_LISTENER);
-                }
-                return;
-            }
             try {
-                if (logger.isDebugEnabled()) {
-                    logger.debug("[{}][{}] creating shard", shardRouting.index(), shardId);
-                }
-                IndexShard indexShard = indexService.createShard(shardRouting);
-                indexShard.addShardFailureCallback(failedShardHandler);
+                logger.debug("{} creating shard", shardRouting.shardId());
+                RecoveryState recoveryState = recoveryState(nodes.getLocalNode(), sourceNode, shardRouting,
+                    indexService.getIndexSettings().getIndexMetaData());
+                indicesService.createShard(shardRouting, recoveryState, recoveryTargetService, new RecoveryListener(shardRouting),
+                    repositoriesService, nodeServicesProvider, failedShardHandler);
             } catch (IndexShardAlreadyExistsException e) {
                 // ignore this, the method call can happen several times
+                assert false;
             } catch (Throwable e) {
-                failAndRemoveShard(shardRouting, indexService, true, "failed to create shard", e);
-                return;
+                failAndRemoveShard(shardRouting, true, "failed to create shard", e);
             }
         }
-        final IndexShard indexShard = indexService.getShard(shardId);
+    }
+
+    private void updateShard(ClusterChangedEvent event, DiscoveryNodes nodes, ShardRouting shardRouting, Shard shard) {
+        final ShardRouting currentRoutingEntry = shard.routingEntry();
+        assert currentRoutingEntry.isSameAllocation(shardRouting) :
+            "local shard has a different allocation id but wasn't cleaning by removeShards. "
+                + "cluster state: " + shardRouting + " local: " + currentRoutingEntry;
 
-        if (indexShard.ignoreRecoveryAttempt()) {
-            // we are already recovering (we can get to this state since the cluster event can happen several
-            // times while we recover)
-            logger.trace("ignoring recovery instruction for shard {} (shard state: [{}])", indexShard.shardId(), indexShard.state());
+        try {
+            shard.updateRoutingEntry(shardRouting, event.state().blocks().disableStatePersistence() == false);
```

Comment:
We can even go one step further and remove the boolean flag completely (as we always pass true).